File: piscine/week2/d01/5/java/src/test/java/unit_tests/BeveragesPricingTest.java
package unit_tests;

import main.java.beverages.*;
import main.java.supplements.*;

import org.junit.Test;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.closeTo;

public class BeveragesPricingTest {
    @Test
    public void computes_coffee_price() {
        Beverage coffee = new Coffee();
        assertThat(coffee.price(), is(closeTo(1.20, 0.001)));
    }

    @Test
    public void computes_tea_price() {
        Beverage tea = new Tea();
        assertThat(tea.price(), is(closeTo(1.50, 0.001)));
    }

    @Test
    public void computes_hot_chocolate_price() {
        Beverage hotChocolate = new HotChocolate();
        assertThat(hotChocolate.price(), is(closeTo(1.45, 0.001)));
    }

    @Test
    public void computes_tea_with_milk_price() {
        Tea tea = new Tea();
        tea.addSupplement(new Milk());
        assertThat(tea.price(), is(closeTo(1.60, 0.001)));
    }

    @Test
    public void computes_coffee_with_milk_price() {
        Coffee coffee = new Coffee();
        coffee.addSupplement(new Milk());
        assertThat(coffee.price(), is(closeTo(1.30, 0.001)));
    }

    @Test
    public void computes_coffee_with_milk_and_cream_price() {
        Coffee coffee = new Coffee();
        coffee.addSupplement(new Milk()).addSupplement(new Cream());
        assertThat(coffee.price(), is(closeTo(1.45, 0.001)));
    }


    @Test
    public void computes_hot_chocolate_with_cream_price() {
        HotChocolate hotChocolate = new HotChocolate();
        hotChocolate.addSupplement(new Cream());
        assertThat(hotChocolate.price(),  is(closeTo(1.60, 0.001)));
    }

    @Test
    public void computes_coffee_with_cinnamon_price() {
        Coffee coffee = new Coffee();
        coffee.addSupplement(new Cinnamon());
        assertThat(coffee.price(), is(closeTo(1.25, 0.001)));
    }
}

File: piscine/week2/d01/5/solutions/scala/fergus_y_pedro/src/test/scala/BeveragesPricingTest.scala
import org.scalatest.FunSuite
import org.scalatest.Matchers._


class BeveragesPricingTest extends FunSuite {

  test("computes coffee price"){
    val coffee = new Coffee
    coffee.price() should be (1.20 +- 0.0001)
  }

  test("computes tea price"){
    val tea = new Tea
    tea.price() should be (1.50 +- 0.0001)
  }

  test("computes hot chocolate price"){
    val hotChocolate = new HotChocolate
    hotChocolate.price() should be (1.45 +- 0.0001)
  }

  test("computes tea with milk price"){
    val teaWithMilk = new TeaWithMilk
    teaWithMilk.price() should be (1.60 +- 0.0001)
  }

  test("computes coffee with milk price"){
    val coffeeWithMilk = new CoffeeWithMilk
    coffeeWithMilk.price() should be (1.30 +- 0.0001)
  }

  test("computes coffee with milk and cream price"){
    val coffeeWithMilkAndCream = new CoffeeWithMilkAndCream
    coffeeWithMilkAndCream.price() should be (1.45 +- 0.0001)
  }

  test("computes hot chocolate with cream price"){
    val hotChocolateWithCream = new HotChocolateWithCream
    hotChocolateWithCream.price() should be (1.60 +- 0.0001)
  }

  test("computes coffee with canela"){
    val coffee: Coffee = new Coffee(List(new Cinnamon))
    coffee.totalPrice() should be (1.25 +- 0.0001)
  }

  test("computes coffee with milk and canela"){
    val coffee: Coffee = new Coffee(List( new Cinnamon, new Milk))
    coffee.totalPrice() should be (1.35 +- 0.0001)
  }

  test("computes coffee with milk, cream and canela"){
    val coffee: Coffee = new Coffee(List( new Cinnamon, new Milk, new Cream))
    coffee.totalPrice() should be (1.50 +- 0.0001)
  }

  test("computes coffee with cream "){
    assertThrows[ExceptionInInitializerError] {
      new Coffee(List( new Cream))
    }
  }
}
